src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -196,13 +243,36 @@ private async Task ProcessQueueAsync()
                         // will be sent back to the client but they can also be captured via HandleNonMutatingRequestError,
                         // though these errors don't put us into a bad state as far as the rest of the queue goes.
                         // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+                        // blocking the request queue for longer periods of time (it enforces parallelizability).
+                        var currentWorkTask = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+
+                        if (CancelInProgressWorkUponMutatingRequest)
+                        {
+                            if (currentWorkCts is null)
+                            {
+                                throw new InvalidOperationException($"unexpected null value for {nameof(currentWorkCts)}");
+                            }
+
+                            if (!concurrentlyExecutingTasks.TryAdd(currentWorkTask, currentWorkCts))
+                            {
+                                throw new InvalidOperationException($"unable to add {currentWorkTask} into {concurrentlyExecutingTasks}");
```

Comment:
Did you mean to use nameof() for those locals? Otherwise I think you're just going to get type names, right?

Reply:
bah, yes!

Reply:
also, just do `Contract.ThrowIfFalse(...TryAdd(...));` :)

Reply:
The Contract class isn't available in this project, from what I've seen, thus all the long winded throwing of InvalidOperationExceptions.

Reply:
Source linked in and changed to use Contract

Reply:
uh oh, test projects are having problems building now because of duplicate Contract/nullability types found in the CommonLanguageServerProtocol assembly and MS.CodeAnalysis.Workspaces.

Reply:
our egneral way to support that is to change the namespace for teh type depending on a #define that a particular project provides.

Reply:
Interesting, glad to hear it's already solved. Can you point me to an example?

Reply:
For example:
![image](https://user-images.githubusercontent.com/4564579/217650087-5aea797c-8c80-47a7-943e-e9601225a3b0.png)

(you'd use a different #if constant here of course).

Reply:
At least for the nullability ones, aren't they required to be in a certain namespace for them to work?

Reply:
what nullability types are you referring to?

Reply:
NotNullAttribute for example, referenced [here](https://github.com/dotnet/roslyn/blob/c2a77e312ec7fb983ef683c9639314b2af70601a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs#L25)

Reply:
looks like you pull in C:\github\roslyn\src\Compilers\Core\Portable\InternalUtilities\NullableAttributes.cs for that.  Which does nothing if you're on netcore (since runtime provides it).

Reply:
I think I'm confusing System.Runtime.CompilerServices.NullableAttribute with the one linked above.